src/providers/MeetingProvider/MeetingManager.ts

```diff
@@ -101,6 +101,18 @@ export class MeetingManager implements AudioVideoObserver {
 
   simulcastEnabled: boolean = false;
 
+  suppressAudioInputDevicePermissionRequest: (() => void) | null = null;
+
+  suppressAudioOutputDevicePermissionRequest: (() => void) | null = null;
+
+  suppressVideoInputDevicePermissionRequest: (() => void) | null = null;
+
+  invokeAudioInputDevicePermissionRequest: (() => void) | null = null;
+
+  invokeAudioOutputDevicePermissionRequest: (() => void) | null = null;
+
+  invokeVideoInputDevicePermissionRequest: (() => void) | null = null;
```

Comment:
This is a very good suggestion, I will change them to an enum. Could you please give me an example here, how should I phrase these function? From my shallow understanding, I find every word is necessary 😅.

The permission only works when permission is unset. If the permission has been already granted before user joins a meeting, they will joins with null devices, and they can change the selected device through the `MeetingControls`.

These function is to help our customer to build the view-only mode easily. But it should be much great if we can have a complete set of mechanisms to manage device permission.

Reply:
I would try to avoid defining these functions if at all possible; they don't mean what they seem to mean. For example, if I call

```
mm.setSuppressAudioInputDevicePermissionRequest(() => {});
mm.chooseAudioInputDevice('default');
```

I might expect there to be no permissions dialog… but there will be one as a side effect of the `getUserMedia` call.

Permissions dialogs are a side-effect of `getUserMedia`, and having been granted permissions is a precondition of `enumerateDevices`. Those are the ground truths of this system.

As such, there is only one thing you need to be able to do: prior to the builder listing devices for the first time, they need to be able to invoke the device label trigger that the device controller ordinarily invokes for you. 

As a builder you might know that you don't need to (_e.g._, because you're running in Electron, or because you have already chosen a device), or you might want control of exactly when.

Just as we will add a `deviceLabels` option to `join`, we can add a single method `invokeDeviceLabelTrigger` that takes `DeviceLabels` as input.

Does that make sense?

Reply:
In either case you'd be setting the device label trigger to a new function — either one that depends on the enum value, or to the exact provided function.

Something a bit like this:

```typescript
  async join({ meetingInfo, attendeeInfo, deviceLabels = DeviceLabels.AudioAndVideo }: MeetingJoinData) {
    …
    await this.initializeMeetingSession(this.configuration, deviceLabels);
  }

  async initializeMeetingSession(
    configuration: MeetingSessionConfiguration,
    deviceLabels: DeviceLabels | DeviceLabelTrigger,
  ): Promise<any> {
    const logger = this.createLogger(configuration);
    const deviceController = new DefaultDeviceController(logger);

    this.meetingSession = new DefaultMeetingSession(
      configuration,
      logger,
      deviceController
    );

    this.audioVideo = this.meetingSession.audioVideo;
    this.setupAudioVideoObservers();

    // Amend `sDLT` accordingly: it now takes an optional trigger argument!
    this.setupDeviceLabelTrigger(deviceLabels);

    // … and now you can do this. It's a bit imprecise, because
    // the builder might want to skip this even if they provide a function,
    // but that's OK.
    if (deviceLabels !== DeviceLabels.None) {
      // TODO: selecting and listing devices should NOT be combined!
      await this.listAndSelectDevices();
    }

    this.publishAudioVideo();
    this.setupActiveSpeakerDetection();
    this.meetingStatus = MeetingStatus.Loading;
    this.publishMeetingStatus();
  }
```